### ISSUE TITLE
[Tutorial]: Wrong example of Our Own Ring-Allreduce

### ISSUE THREAD

**Issue by Lyken17** [pytorch/pytorch#31572]
Current, [Our Own Ring-Allreduce](https://pytorch.org/tutorials/intermediate/dist_tuto.html#our-own-ring-allreduce) example is not correct. It will NOT perform all reduce on selected tensors. The code to reproduce is attached below

 ```python
""" Implementation of a ring-reduce with addition. """
def allreduce(send, recv):
    rank = dist.get_rank()
    size = dist.get_world_size()
    send_buff = th.zeros(send.size())
    recv_buff = th.zeros(send.size())
    accum = th.zeros(send.size())
    accum[:] = send[:]

    left = ((rank - 1) + size) % size
    right = (rank + 1) % size

    for i in range(size - 1):
        if i % 2 == 0:
            # Send send_buff
            send_req = dist.isend(send_buff, right)
            dist.recv(recv_buff, left)
            accum[:] += recv[:]
        else:
            # Send recv_buff
            send_req = dist.isend(recv_buff, right)
            dist.recv(send_buff, left)
            accum[:] += send[:]
        send_req.wait()
    recv[:] = accum[:]

def run_allreduce(rank, size):
    data = torch.ones(3)  * (rank + 1)    
    recv = torch.zeros_like(a)
    allreduce(send=data, recv=recv)
    print(recv)

def init_process(rank, size, fn, backend='gloo'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group(backend, rank=rank, world_size=size)
    fn(rank, size)


if __name__ == "__main__":
    size = 4
    processes = []
    for rank in range(size):
        p = Process(target=init_process, args=(rank, size, run_allreduce))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

The terminal output is 

```bash
~/Workspace$ python main.py
# original data
Message from 0 tensor([0., 0., 0.])
Message from 1 tensor([1., 1., 1.])
# data after allreduce (does not work as expect)
Message from 0 tensor([0., 0., 0.])
Message from 1 tensor([1., 1., 1.])
```

The correct code snippet should be 

```python
def allreduce(send, recv):
    rank = dist.get_rank()
    size = dist.get_world_size()
    send_buff = send.clone()
    recv_buff = send.clone()
    accum = send.clone()

    left = ((rank - 1) + size) % size
    right = (rank + 1) % size

    for i in range(size - 1):
        if i % 2 == 0:
            # Send send_buff
            send_req = dist.isend(send_buff, right)
            dist.recv(recv_buff, left)
            accum[:] += recv_buff[:]
        else:
            # Send recv_buff
            send_req = dist.isend(recv_buff, right)
            dist.recv(send_buff, left)
            accum[:] += send_buff[:]
        send_req.wait()
    recv[:] = accum[:]
```

Though I didn't quite understand the usage of `send_buffer` and `recv_buffer` here, to avoid read/write collision? 

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

**Comment by rohan-varma** [comment]
Thanks for pointing this out! The reason for using `send_buffer` and `recv_buffer` is so we don't overwrite the `send` tensor with the data that we receive.

**Comment by Lyken17** [comment]
@rohan-varma How about the following code? I think it provides the same functionality and is easier to understand.
```python
send_req = dist.isend(send_buff, right)
dist.recv(recv_buff, left)  # recv is a blocking operation.
accum[:] += recv_buff[:]
send_buff[:] = recv_buff[:]
```

**Comment by gchanan** [comment]
@rohan-varma is this high priority?  Can you add the label if you think it is?

**Comment by rohan-varma** [comment]
@gchanan I've added the high priority label since this is an error in correctness of our docs that we should prioritize fixing.

**Comment by mattip** [comment]
This is an issue in the pytorch/tutorial repo

**Comment by Lyken17** [comment]
So, maybe I should propose a PR? It has been a while since it was reported.

**Comment by mrshenli** [comment]
Hey @Lyken17, a PR would be highly appreciated. Thanks!

cc @rohan-varma

**Comment by Antymon** [comment]
I wanted to report an issue on the very same thing. Half a year after the original posting this keeps confusing people like me.

**Comment by rohan-varma** [comment]
@Lyken17, was wondering if you're still planning on sending a PR for this? Thanks!

**Comment by Lyken17** [comment]
@rohan-varma Oh, too many other tasks recently. Forget this issue. 

Sure, I can make a PR. Submit to https://github.com/pytorch/tutorials/blob/master/intermediate_source/dist_tuto.rst?

**Comment by rohan-varma** [comment]
> Sure, I can make a PR. Submit to https://github.com/pytorch/tutorials/blob/master/intermediate_source/dist_tuto.rst?

Thank you! Yes, feel free to send a PR to that file over at the [pytorch/tutorials](https://github.com/pytorch/tutorials) repo and add me as a reviewer.

**Comment by Lyken17** [comment]
@rohan-varma https://github.com/pytorch/tutorials/pull/1039 I've created one, please have a look.

**Comment by Lyken17** [comment]
As the fix is now merged into tutorial master branch, I think we can close this issue now. Feel free to reopen if you have additional comments.
